sdk/src/router/public/router-utils.ts

```diff
@@ -0,0 +1,217 @@
+  /**
+   * Selects the best executable route from a list of routes using the current execution environment.
+   * The wallet support type, available ATA accounts, existance of lookup tables all effect the transaction size
+   * and eligibility of a route.
+   *
+   * @param ctx The {@link WhirlpoolContext} that represents the current execution environment
+   * @param orderedRoutes A list of routes to select from, ordered by the best routes (trade amount wise) first.
+   * @param opts {@link RouteSelectOptions} to configure the selection of the best route.
+   * @returns
+   * The best {@link ExecutableRoute} that can be used to execute a swap. If no executable route is found, null is returned.
+   */
+  static async selectFirstExecutableRoute(
+    ctx: WhirlpoolContext,
+    orderedRoutes: TradeRoute[],
+    opts: RouteSelectOptions
+  ): Promise<ExecutableRoute | null> {
+    const { wallet } = ctx;
+
+    if (orderedRoutes.length === 0) {
+      return null;
+    }
+
+    // Don't measure if there is no wallet
+    if (!isWalletConnected(wallet)) {
+      return [orderedRoutes[0], undefined];
+    }
+
+    // Preload LookupTableFetcher with lookup tables that are needed for v0 transactions
+    if (opts.maxSupportedTransactionVersion !== "legacy" && ctx.lookupTableFetcher) {
+      await loadLookupTablesForRoutes(ctx.lookupTableFetcher, orderedRoutes);
+    }
+
+    for (let i = 0; i < orderedRoutes.length && i < MEASURE_ROUTE_MAX; i++) {
+      const route = orderedRoutes[i];
+
+      const tx = await getSwapFromRoute(
+        ctx,
+        {
+          route,
+          slippage: Percentage.fromFraction(0, 100),
+          resolvedAtaAccounts: opts.availableAtaAccounts ?? null,
+          wallet: wallet.publicKey,
+        },
+        false
+      );
+
+      if (!!opts.onRouteEvaluation) {
+        opts.onRouteEvaluation(route, tx);
+      }
+
+      try {
+        const legacyTxSize = await tx.txnSize({
+          latestBlockhash: MEASUREMENT_BLOCKHASH,
+          maxSupportedTransactionVersion: "legacy",
+        });
+        if (legacyTxSize !== undefined && legacyTxSize < TX_LIMIT) {
+          return [route, undefined];
+        }
+      } catch (e) {
+        // No-op
+      }
+
+      let v0TxSize;
+      if (opts.maxSupportedTransactionVersion !== "legacy" && ctx.lookupTableFetcher) {
+        const addressesToLookup = RouterUtils.getTouchedTickArraysFromRoute(route);
+        if (addressesToLookup.length > MAX_LOOKUP_TABLE_FETCH_SIZE) {
+          continue;
+        }
+
+        const lookupTableAccounts = await ctx.lookupTableFetcher.getLookupTableAccountsForAddresses(
+          addressesToLookup
+        );
+        try {
+          v0TxSize = await tx.txnSize({
+            latestBlockhash: MEASUREMENT_BLOCKHASH,
+            maxSupportedTransactionVersion: opts.maxSupportedTransactionVersion,
+            lookupTableAccounts,
+          });
+
+          if (v0TxSize !== undefined && v0TxSize < ENCODED_LIMIT) {
```

Comment:
I think we need to use 1 constant for size comparison.

Now ``TransactionBuilder.txnSize`` returns the size of Base64 encoded data, so we need to use ``ENCODED_LIMIT`` for both legacy and v0.

But in my opinion, it is better to update ``TransactionBuilder.txnSize`` to return the size of UNencoded data, and use ``TX_LIMIT`` (or ``PACKET_DATA_SIZE`` from web3.js) for both legacy and v0.

Reply:
PR https://github.com/orca-so/whirlpools/pull/100 will fix this. 🙂